python/src/iceberg/io/base.py

```diff
@@ -24,7 +24,40 @@
 """
 
 from abc import ABC, abstractmethod
-from typing import Union
+from typing import Protocol, Union, runtime_checkable
+
+
+@runtime_checkable
+class InputStream(Protocol):
+    def read(self, n: int) -> bytes:
+        ...
```

Comment:
docs for methods?  Or maybe a top level doc that these should have the same semantics as IOBase?

Reply:
Are these copied from `IOBase`? I'm not sure what should be here for Python. Are these the methods needed for Arrow to read from this stream?

Reply:
These are the equivalent of [SeekableInputStream](https://github.com/apache/iceberg/blob/master/python_legacy/iceberg/api/io/seekable_input_stream.py#L19) and [PositionOutputStream](https://github.com/apache/iceberg/blob/master/python_legacy/iceberg/api/io/position_output_stream.py#L19) in the legacy python client. It's just less strict here in that any class with the same structure is automatically a subclass.

I think this highlights the behavior:
```py
from typing import Protocol, runtime_checkable

@runtime_checkable
class A(Protocol):
    def foo(self):
        ...
    def bar(self):
        ...

class B:
    def foo(self):
        ...
    def bar(self):
        ...

class C:
    def foo(self):
        ...
    def bar(self):
        ...
    def baz(self):
        ...

class D:
    def foo(self):
        ...

isinstance(B(), A)  # True
isinstance(C(), A)  # True
isinstance(D(), A)  # False
```

So an input file or output file class could have more methods, as long as it has the required methods set in the protocol which should be the case for any file-like object that's based on [RawIOBase](https://docs.python.org/3/library/io.html#io.RawIOBase) (inherits from IOBase and adds a `read`, `readall`, `readinto`, and `write` method).

The reason I renamed a few of the methods from the legacy client was also to be in-line with `RawIOBase`. Here's an example of how urllibs HTTPResponse object matches this protocol:
```py
import urllib.request

response = urllib.request.urlopen('http://python.org/')
isinstance(response, InputStream)  # True

print(dir(response))
```
output:
```
[...
 'begin',
 'chunk_left',
 'chunked',
 'close',
 'closed',
 'code',
 'debuglevel',
 'detach',
 'fileno',
 'flush',
 'fp',
 'getcode',
 'getheader',
 'getheaders',
 'geturl',
 'headers',
 'info',
 'isatty',
 'isclosed',
 'length',
 'msg',
 'peek',
 'read',
 'read1',
 'readable',
 'readinto',
 'readinto1',
 'readline',
 'readlines',
 'reason',
 'seek',
 'seekable',
 'status',
 'tell',
 'truncate',
 'url',
 'version',
 'will_close',
 'writable',
 'write',
 'writelines']
```

One of the remaining questions though is where do we set this `isinstance` check, or do we do it at all? The rest of the library will be assuming this interface so do we leave it to every `InputFile` and `OutputFile` implementations to do an `isinstance` check? Or do we do the `isinstance` check in other parts of the library wherever `FileIO.new_input_file(...)` is called, for example. This feels like it might be a tradeoff to this approach.

(sorry for the super long comment)

Reply:
I'm much less concerned about running an isinstance check than about whether anything that conforms to this protocol can be passed into Arrow to read a Parquet file, or the equivalent on the write path. I'm not familiar with that requirement... so how do you know that this is going to work and we don't require more methods?

Reply:
Ah I see. [This](https://arrow.apache.org/docs/python/parquet.html#reading-and-writing-single-files) section of the pyarrow docs suggests that an instance of [pyarrow.NativeFile](https://arrow.apache.org/docs/python/generated/pyarrow.NativeFile.html#pyarrow.NativeFile) will provide the best performance. It has many more methods but includes the methods defined in both protocols here so it would be implicitly considered a subtype of both.

Reply:
Sounds like NativeFile could be returned here, which is good. I think IOBase could be as well. But I'm wondering what the requirement is for what is necessary for passing this IN to Arrow?

Reply:
I don't think NativeFile is the right thing to return.  NativeFile is arrow's file abstraction and has native C++ implementations for the most part.  [PythonFile](https://arrow.apache.org/docs/python/generated/pyarrow.PythonFile.html?highlight=pythonfile#pyarrow.PythonFile) is the adapter from File-like python objects to Arrow's File format.  I think IOBase probably guarantees this and looking at the [source](https://github.com/apache/arrow/blob/e9e16c9da7a76718640f2b3f23200a3755790011/python/pyarrow/io.pxi#L679) it seems like there are some isinstance checks for IOBase or duck-typed.  

What we probably want to do in whatever code passes these interfaces to Arrow is to check if they are an instance of a wrapper around NativeFile and pass the native files instead to avoid GIL and non-zero indirection costs as mentioned in the PythonFile docs.

Reply:
@emkornfield I think I follow. It sounds like when using FileIO instances throughout the library, you're saying we have to support those that produce file-like objects (which we pass into `PythonFile` then into pyarrow) as well as those that produce `pyarrow.NativeFile` instances which can just be passed directly to pyarrow.

In that case, instead of defining Protocols, should we just use typehints for the `InputFile` and `OutputFile` classes to specify that either [typing.IO](https://docs.python.org/3/library/typing.html#typing.IO) or `pyarrow.NativeFile` can be returned?
```py
from abc import ABC, abstractmethod
from typing import Union, IO

from pyarrow import NativeFile

class InputFile(ABC):
  ...
  def open(self) -> Union[IO, NativeFile]
    ...

class OutputFile(ABC):
  ...
  def create(self) -> Union[IO, NativeFile]:
    ...
```

Reply:
So arrow should be [smart enough](https://arrow.apache.org/docs/python/parquet.html#reading-and-writing-single-files) to distinguish between NativeFile and IOBase and it appears NativeFile is protocol compatible with IOBase.  The issues I see with the UnionType are:
1.  It appears IO is not a protocol, so I think type checks would fail unless consumers specifically extended it (I haven't done a lot of development with type hinting in python though so I'm not sure).
2.  It seems like it takes a hard dependency on pyarrow, which in previous discussions we wanted to avoid. 

Given this I still think using a protocol that mimics IO[bytes] seems like what we would want and for an arrow specific implementation should avoid any performance loss.

Reply:
I agree that the protocol is the right approach. You should be able to use anything, as long as Arrow can use it to read or write a file. The question is then what that protocol needs to validate. I think for that we just need to try it out until we have a minimal set of functions. Or maybe an Arrow expert can help us out.

Reply:
TL;DR; My best guess is (that take bytes):
- `read(...)`
- `write(...)`
- `seek(...)`
- `tell(...)`
- `writable()`
-  `close()`
- `closed()`

I don't know if other libraries (e.g. json) might require more.

I think the safe thing is to use IOBase methods as the protocol.  The code linked above mostly seems to only require "writeable()".  The C++ code that does the adaptation is [here](https://github.com/apache/arrow/blob/8e43f23dcc6a9e630516228f110c48b64d13cec6/cpp/src/arrow/python/io.cc) it looks like it accesses most IOBase methods and `read(...)`/`write(...)` methods (these I think should take there signatures from [RawIOBase](https://docs.python.org/3/library/io.html#io.RawIOBase).  The methods that don't look like they are necessary are 'isatty` and `fileno`.  Some of the other methods are just straight pass-through to the object (`truncate`, `readline`, `readlines`) but even if Arrow won't blow-up I think the question is what other libraries to we expect to interact with and if those require the methods.

Reply:
> I think for that we just need to try it out until we have a minimal set of functions.

Great idea @rdblue, I tried this out. @emkornfield it looks like your guess was right--although `writable` was the only method that could be removed without causing any error. I found that interesting since the `PythonFile` source code you linked to seems to need that to [infer the handler mode](https://github.com/apache/arrow/blob/e9e16c9da7a76718640f2b3f23200a3755790011/python/pyarrow/io.pxi#L703). I might be missing where that's inferred some other way, maybe duck-typing.

Read requires:
- read(...)
- seek(...)
- tell(...)
- closed()
- close()

Write requires:
- write(...)
- closed()
- close()

This was the code I used to test this out:
```py
from pyarrow import parquet as pq


class InputFileImpl:
    def __init__(self, filepath):
        file_object = open(filepath, "rb")

        # Required methods for use in pq.read_table(...)
        self.read = file_object.read
        self.seek = file_object.seek
        self.tell = file_object.tell
        self.closed = file_object.closed
        self.close = file_object.close

class OutputFileImpl:
    def __init__(self, filepath):
        file_object = open(filepath, "wb")

        # Required methods for use in pq.write_table(...)
        self.write = file_object.write
        self.closed = file_object.closed
        self.close = file_object.close


input_file = InputFileImpl("example.parquet")
table = pq.read_table(input_file)

output_file = OutputFileImpl("example_output.parquet")
pq.write_table(table, output_file)
output_file.close()
table_reread = pq.read_table("example_output.parquet")

table.equals(table_reread)  # returns True
```

Reply:
I'm a bit surprised that `tell` was required for read, but not for write... in the JVM implementation we only need the position in the write path, not the read path. Maybe that's tracked internally by the Arrow Parquet writer, though. (@xhochy might know)

Reply:
I'd guess for parquet the `mode` is passed through as a parameter so that code path isn't hit.

Tell is used in determining file size I believe on the read path (capturing and restoring current position).  For write path I think parquet does a fair amount of internal state tracking.  This could be an artifact of the file abstractions in parquet being initially separate from Arrow

Reply:
I updated this PR to only include the minimally required pyarrow methods for the protocols. I also dropped 3.7 from the tox configuration and python-ci workflow.